### PR TITLE
PixelCal before Tweak Peak Peek

### PR DESCRIPTION
**Pull request by rboston628**
## Description of work

This performs pixel calibration prior to adjustments of peak lists going into group calibration.

## Explanation of work

Now that the recipes and service endpoints have been split, this is able to call pixel calibration before the tweak peak stage, so that it only needs to occur once.

### Dev testing

Turn on CIS mode.

Run calibration in GUI using 46680 and Bank.  Do NOT skip pixel calibration.  At the end, and error box will tell you to skip pixel calibration.

Do NOT switch the toggle yet.

Make sure the cross-correlation workspaces are present and examine them.  They are the following:
- _46680_tmp_group_CC_0_group1
- _46680_tmp_group_CC_0_group2
- _46680_tmp_group_offset_0
- dsp_046680_raw_pixel0
- dsp_046680_raw_startOfPixelDiffCal
- offsets_46680_0

This proves that pixel calibration is being called immediately following the request screen.  Remove all of these workspaces.

Switch the skip toggle, then continue.  Observe that it goes directly to tweak peak without performing pixel calibration.  There should be no cross correlation workspaces.  The logs should not show any calls to CrossCorrelation or GetDetectorOffsets.

At tweak peak, the params you need are
- grouping: Bank
- FWHM Left: 2
- FWHM Right: 5

Continue all the way to saving.

It would be good to check with a better run, like 58882 and Silicon, to make sure the Purge Bad Peaks behavior is still working as intended.

**UPDATE**

Close SNAPRed.  Open it again to the diffcal panel.  With the workspace list *completely blank* select to skip pixel calibration.  Use any run.  Make sure this completes normally with no issues.

### CIS testing

See the new `reorder_proto.py` CIS script.  This shows how the data looks after pixel calibrating, focusing, and fitting, which are the steps that happen prior to group calibration.  Make sure this is working as intended.

Then run the SNAPRed GUI with the same data and make sure it is also working as intended, and as in the prototype.

## Link to EWM item

[EWM#7387](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=7387)

[EWM#7617](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=7617)

### Verification

- [x] the author has read the EWM story and acceptance critera
- [x] the reviewer has read the EWM story and acceptance criteria
- [ ] the reviewer certifies the acceptance criteria below reflect the criteria in EWM

### Acceptance Criteria

This list is for ease of reference, and does not replace reading the EWM story as part of the review.  Verify this list matches the EWM story before reviewing.

- [ ] PixelCalibration executes before peak adjustment stage
- [ ] Between Pixel and Group calibrations, the Workflow displays a screen with the peaks after the pixel calibration has been applied, allowing user to adjust parameters before running Group Diffraction Calibration
- [ ] After adjustment, workflow continues to GroupCalibration then to Save

### CODE REVIEW

File: tests/unit/backend/recipe/test_DiffractionCalibrationRecipe.py
Comment (rboston628) [inline tests/unit/backend/recipe/test_DiffractionCalibrationRecipe.py:1]: Note: all of these tests were reproduced in `test_CalibrationService`

**Comment by darshdinger** [comment]
## Testing:
---------------
1. Trying DiffCal workflow with suggested inputs:
image

2. Workspace present:
image

3. Clear workspaces and toggle the Skip Pixel Calibration:
image

4. Correct the peaks using given FWHM parameters:
image

5. Complete Workflow:
Got this error - 
image


6. Trying with 58882:
Before Purge Bad Peaks -
image
After Purge Bad Peaks -
image
Same error as before -
image

**Comment by rboston628** [comment]
The error @darshdinger reported above is related to [EWM 7617](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=7617).

@walshmm  had done some fixes to it in PR #486.

This PR borrowed some of those fixes to solve the same defect.

**Comment by darshdinger** [comment]
## Testing Fixed Branch
----------------------------------
1. Start DiffCal with stateless local directory tree:
image

2. Run DiffCal without Skip Pixel Calibration:
image

3. Enable Skip Pixel Calibration:
image

**Note:** Purge Bad Peaks functions as it should. Able to successfully complete workflow.

**Review by darshdinger — APPROVED** [review]
Tested fixes, this PR is functional and meets the acceptance criteria.

**Pull request merged**